library/std/src/sys/windows/compat.rs

```diff
@@ -49,81 +49,211 @@
 //! * call any Rust function or CRT function that touches any static
 //!   (global) state.
 
-macro_rules! compat_fn {
-    ($module:literal: $(
+use crate::ffi::{c_void, CStr};
+use crate::ptr::NonNull;
+use crate::sys::c;
+
+/// Helper macro for creating CStrs from literals and symbol names.
+macro_rules! ansi_str {
+    (sym $ident:ident) => {{
+        #[allow(unused_unsafe)]
+        unsafe {
+            CStr::from_bytes_with_nul_unchecked(concat!(stringify!($ident), "\0").as_bytes())
+        }
+    }};
+    ($lit:literal) => {{
+        #[allow(unused_unsafe)]
+        unsafe {
+            CStr::from_bytes_with_nul_unchecked(concat!($lit, "\0").as_bytes())
+        }
+    }};
+}
+
+#[used]
+#[link_section = ".CRT$XCU"]
+static INIT_TABLE_ENTRY: unsafe extern "C" fn() = init;
+
+/// This is where the magic preloading of symbols happens.
+///
+/// We don't preload all lazily loaded functions, only those which don't have
+/// a fallback or may be particularly perf sensitive.
+unsafe extern "C" fn init() {
+    // There is no locking here. This code is executed before main() is entered, and
+    // is guaranteed to be single-threaded.
+    //
+    // DO NOT do anything interesting or complicated in this function! DO NOT call
+    // any Rust functions or CRT functions if those functions touch any global state,
+    // because this function runs during global initialization. For example, DO NOT
+    // do any dynamic allocation, don't call LoadLibrary, etc.
+
+    if let Some(synch) = Module::new(c::SYNCH_API) {
+        // These are optional and so we must manually attempt to load them
+        // before they can be used.
+        c::WaitOnAddress::preload(synch);
+        c::WakeByAddressSingle::preload(synch);
+    }
+
+    if let Some(kernel32) = Module::new(c::KERNEL32) {
+        // Preloading this means getting a precise time will be as fast as possible.
+        c::GetSystemTimePreciseAsFileTime::preload(kernel32);
+    }
+
+    if let Some(ntdll) = Module::new(c::NTDLL) {
+        // A program is almost certainly going to be doing some kind of I/O.
+        c::NtReadFile::preload(ntdll);
+        c::NtWriteFile::preload(ntdll);
+        c::RtlNtStatusToDosError::preload(ntdll);
+    }
+}
+
+/// Represents a loaded module.
+///
+/// Note that the modules std depends on must not be unloaded.
+/// Therefore a `Module` is always valid for the lifetime of std.
+#[derive(Copy, Clone)]
+pub(in crate::sys) struct Module(NonNull<c_void>);
+impl Module {
+    /// Try to get a handle to a loaded module.
+    pub fn new(name: &CStr) -> Option<Self> {
```

Comment:
It seems like this ought to be unsafe (or proc_address), since based on the docs it sounds like we don't really have any guarantees the pointer will remain valid, particularly later in the program when there's other threads.

Reply:
Hm, I guess there are no guarantees but my thinking was that it's freeing the library that would be unsafe rather than getting a handle. The handle will remain valid for as long as the module remains loaded and the stdlib absolutely requires that the DLLs it depends on aren't unloaded, otherwise calling into them would be UB.

Reply:
Ok, on reflection I've made `new()` unsafe and added a comment about why.